api/src/service/domain/organization/user_password_change.spec.ts

```diff
@@ -0,0 +1,102 @@
+import { changeUserPassword, RequestData } from "./user_password_change";
+import { UserRecord } from "./user_record";
+
+const ctx: Ctx = { requestId: "", source: "test" };
+const root: ServiceUser = { id: "root", groups: [] };
+const alice: ServiceUser = { id: "alice", groups: ["alice_and_bob", "alice_and_bob_and_charlie"] };
+const bob: ServiceUser = { id: "bob", groups: ["alice_and_bob", "alice_and_bob_and_charlie"] };
+
+const dummy = "dummy";
+const passwordChangeUser: UserRecord = {
+  id: dummy,
+  createdAt: new Date().toISOString(),
+  displayName: dummy,
+  organization: dummy,
+  passwordHash: dummy,
+  address: dummy,
+  encryptedPrivKey: dummy,
+  permissions: {},
+  log: [],
+  additionalData: {},
+};
+
+const requestData: RequestData = {
+  userId: dummy,
+  newPassword: "newtest",
+};
+
+const baseRepository = {
+  getUser: () => Promise.resolve(passwordChangeUser),
+  hash: () => Promise.resolve("passwordHash"),
+};
+
+describe("change a user's password: authorization", () => {
+  it("Without the user.changePassword permission, a user cannot change a password", async () => {
+    const result = await changeUserPassword(ctx, alice, requestData, {
+      ...baseRepository,
+    });
+    assert.instanceOf(result, NotAuthorized);
+  });
+
+  it("The root user doesn't need permission to change a user's password", async () => {
+    const result = await changeUserPassword(ctx, root, requestData, {
+      ...baseRepository,
+    });
+    assert.isTrue(Result.isOk(result), (result as Error).message);
+  });
+
+  it("A user can change another user's password if the correct permissions are given", async () => {
+    const result = await changeUserPassword(ctx, alice, requestData, {
+      ...baseRepository,
+      getUser: () =>
+        Promise.resolve({
+          ...passwordChangeUser,
+          permissions: { "user.changePassword": [alice.id] },
+        }),
+    });
+    assert.isTrue(Result.isOk(result), (result as Error).message);
+  });
+});
+
+describe("change a user's password: how modifications are applied", () => {
+  it("Changes the user's password immediately", async () => {
+    // Set password hash
+    // select new password
+    // check if new hash is still valid
+    const oldPassword = "passwordTest";
+    const oldPasswordHash = await hashPassword(oldPassword);
+    const reqData = {
+      userId: dummy,
+      newPassword: "newPassword",
+    };
+
+    assert.isTrue(await isPasswordMatch(oldPassword, oldPasswordHash));
+    const result = await changeUserPassword(ctx, alice, reqData, {
+      ...baseRepository,
+      hash: async passwordPlaintext => hashPassword(passwordPlaintext),
+      getUser: async () =>
+        Promise.resolve({
+          ...passwordChangeUser,
+          permissions: { "user.changePassword": [alice.id] },
+        }),
+    });
+    if (Result.isErr(result)) {
+      throw result;
+    }
+
+    const sourcedUser = newUserFromEvent(ctx, passwordChangeUser, result[0]);
```

Comment:
use the sourceUsers function here and pass the full list of events. or use reduce like this:

```typescript
const sourcedUser = result.reduce((user, event) => newUserFromEvent(ctx, user, event), passwordChangeUser);
```